src/value.rs

```diff
@@ -960,27 +960,57 @@ impl Value {
 
         let mut stack = vec![State::ProcessType(ty)];
         let mut result_stack = vec![];
-        while let Some(state) = stack.pop() {
+        'stack_loop: while let Some(state) = stack.pop() {
             match state {
-                State::ProcessType(ty) => match ty.bound() {
-                    CompleteBound::Unit => result_stack.push(Value::unit()),
-                    CompleteBound::Sum(ref l, ref r) => {
-                        if !bits.next().ok_or(EarlyEndOfStreamError)? {
-                            P::read_left_padding(bits, l, r)?;
-                            stack.push(State::DoSumL(Arc::clone(r)));
-                            stack.push(State::ProcessType(l));
-                        } else {
-                            P::read_right_padding(bits, l, r)?;
-                            stack.push(State::DoSumR(Arc::clone(l)));
-                            stack.push(State::ProcessType(r));
+                State::ProcessType(ty) if ty.tmr() == Tmr::POWERS_OF_TWO[0] => {
+                    result_stack.push(Value::u1(bits.read_bit()?.into()));
+                }
+                State::ProcessType(ty) if ty.tmr() == Tmr::POWERS_OF_TWO[1] => {
+                    result_stack.push(Value::u2(bits.read_u2()?.into()));
+                }
+                State::ProcessType(ty) if ty.tmr() == Tmr::POWERS_OF_TWO[2] => {
+                    let u4 = (u8::from(bits.read_u2()?) << 2) + u8::from(bits.read_u2()?);
+                    result_stack.push(Value::u4(u4));
+                }
+                State::ProcessType(ty) => {
+                    // The POWERS_OF_TWO array is somewhat misnamed; the ith index contains
+                    // the TMR of TWO^(2^n). So e.g. the 0th index is 2 (a bit), the 1st is
+                    // u2, then u4, and the 3rd is u8.
```

Comment:
987a8a20fadc1334a19a649bc458b6543095fb6a: This comment should be in the documentation of `POWERS_OF_TWO` / we should rename `POWERS_OF_TWO` in a follow-up commit.